modules/rostests/apitests/shell32/SHCreateFileDataObject.cpp

```diff
@@ -0,0 +1,458 @@
+/*
+ * PROJECT:     ReactOS api tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Test for SHCreateFileDataObject
+ * COPYRIGHT:   Copyright 2019-2021 Mark Jansen <mark.jansen@reactos.org>
+ *
+ * This is 99% the same as the test for SHCreateDataObject, except that this always has 4 data types (TestDefaultFormat)
+ */
+
+#include "shelltest.h"
+#include <ndk/rtlfuncs.h>
+#include <stdio.h>
+#include <shellutils.h>
+#include <shlwapi.h>
+
+static DWORD g_WinVersion;
+
+typedef HRESULT(WINAPI* tSHCreateFileDataObject)(PCIDLIST_ABSOLUTE pidlFolder, UINT cidl, PCUITEMID_CHILD_ARRAY apidl, IDataObject* pDataInner, IDataObject** ppDataObj);
+static tSHCreateFileDataObject pSHCreateFileDataObject;
+
+
+static void TestAdviseAndCanonical(PCIDLIST_ABSOLUTE pidlFolder, UINT cidl, PCUIDLIST_RELATIVE_ARRAY apidl)
+{
+    CComPtr<IDataObject> spDataObj;
+    HRESULT hr = pSHCreateFileDataObject(pidlFolder, cidl, apidl, NULL, &spDataObj);
+
+    ok_hex(hr, S_OK);
+    if (!SUCCEEDED(hr))
+        return;
+
+    hr = spDataObj->DAdvise(NULL, 0, NULL, NULL);
+    ok_hex(hr, OLE_E_ADVISENOTSUPPORTED);
+
+    hr = spDataObj->DUnadvise(0);
+    ok_hex(hr, OLE_E_ADVISENOTSUPPORTED);
+
+    hr = spDataObj->EnumDAdvise(NULL);
+    ok_hex(hr, OLE_E_ADVISENOTSUPPORTED);
+
+
+    FORMATETC in = {1, (DVTARGETDEVICE*)2, 3, 4, 5};
+    FORMATETC out = {6, (DVTARGETDEVICE*)7, 8, 9, 10};
+
+    hr = spDataObj->GetCanonicalFormatEtc(&in, &out);
+    ok_hex(hr, DATA_S_SAMEFORMATETC);
+
+    if (g_WinVersion < _WIN32_WINNT_VISTA)
+    {
+        ok_int(out.cfFormat, 6);
+        ok_ptr(out.ptd, (void*)7);
+        ok_int(out.dwAspect, 8);
+        ok_int(out.lindex, 9);
+        ok_int(out.tymed, 10);
+        trace("out unmodified\n");
+    }
+    else
+    {
+        ok_int(out.cfFormat, in.cfFormat);
+        ok_ptr(out.ptd, NULL);
+        ok_int(out.dwAspect, (int)in.dwAspect);
+        ok_int(out.lindex, in.lindex);
+        ok_int(out.tymed, (int)in.tymed);
+        trace("in copied to out\n");
+    }
+}
+
+
+#define ok_wstri(x, y) \
```

Comment:
This macro seems to be redefined many times in this apitest/shell32... maybe consider putting it in a global header of this apitest?

Reply:
Feel free to, after this PR is merged.